packages/core/src/browser/widgets/widget.ts

```diff
@@ -45,7 +45,17 @@ export class BaseWidget extends Widget {
         this.dispose();
     }
 
+    onBeforeAttach(msg: Message): void {
+        if (!this.hasIconInTitle()) {
+            this.addNoIconClassToTitle();
+        }
+        super.onBeforeAttach(msg);
+    }
```

Comment:
Do we really need to make the `no-icon` class handling extensible? IMO this could be hard-coded here, e.g.
```typescript
    protected onBeforeAttach(msg: Message): void {
        if (this.title.iconClass === '') {
            this.title.iconClass = 'no-icon';
        }
        super.onBeforeAttach(msg);
    }
```

Reply:
Yes we could simply do 
```
        if (this.title.iconClass === '') {
            this.title.iconClass = 'no-icon';
        }
```
in the onBeforeAttach(), and 

```
        if (this.title.iconClass.includes('no-icon')) {
            this.title.iconClass = '';
        }
```

in the onBeforeDetach().

If anyone touches

Reply:
👍 

To be symmetric, the cleanup code should go to `onAfterDetach`, and the check should be `this.title.iconClass === 'no-icon'` instead of using `includes`. Also remember to make those methods `protected` like in the `Widget` superclass.

Reply:
done. thanks !